### DISCUSSION TITLE
How can I use mongoose.Types.DocumentArray with subdocument?

### DISCUSSION THREAD

**Discussion by xkguq007** [typegoose/typegoose#380] [Q&A] [answered]
## Versions

- typegoose: 7.3.2
- mongoose: 5.9.29

Sorry for bad English.

What I want to use some mongoose's subdocument functions such as 'id', and I want to use _id as ObjectId at the same time, without any explicit conversion.

Below is what I've done.
```ts
export class DeliveryAddressItem extends defaultClasses.Base {
  @prop({ required: true })
  public name!: string;
}

export interface DeliveryAddress extends defaultClasses.Base {}
export class DeliveryAddress extends defaultClasses.TimeStamps {
  @prop({ required: true, type: () => DeliveryAddressItem, default: [] })
  public items!: mongoose.Types.DocumentArray< DeliveryAddressItem >;
}
```
Because I want to use subdocument function 'id'
```ts
deliveryAddress.items.id('object id');
```
So, DeliveryAddress has items which has _id (subdocument, not nested array).

Then, there is following error.
<img width="502" alt="image" src="https://user-images.githubusercontent.com/9859504/93839361-b09d1f00-fcc7-11ea-906b-0efbc7a4913f.png">

However above error can solve following code, just change Base to MongooseDocument for subdocument class.
```ts
export class DeliveryAddressItem extends mongoose.MongooseDocument {
  @prop({ required: true })
  public name!: string;
}

export interface DeliveryAddress extends defaultClasses.Base {}
export class DeliveryAddress extends defaultClasses.TimeStamps {
  @prop({ required: true, type: () => DeliveryAddressItem, default: [] })
  public items!: mongoose.Types.DocumentArray< DeliveryAddressItem >;
}
```

Then, I MUST USE _id as any WITHOUT (_id as mongoose.Types.ObjectId); // I don't want to explicit conversion.

So, I try following codes.
```ts
export interface DeliveryAddressItem extends defaultClasses.Base {}
export class DeliveryAddressItem extends mongoose.MongooseDocument {
  @prop({ required: true })
  public name!: string;
}

export interface DeliveryAddress extends defaultClasses.Base {}
export class DeliveryAddress extends defaultClasses.TimeStamps {
  @prop({ required: true, type: () => DeliveryAddressItem, default: [] })
  public items!: mongoose.Types.DocumentArray< DeliveryAddressItem >;
}
```

Then, there is following error.
<img width="497" alt="image" src="https://user-images.githubusercontent.com/9859504/93839775-f60e1c00-fcc8-11ea-8feb-71a0e0e59ddd.png">

How can I solve this?

**Comment by hasezoey** [comment, marked as answer]
i think you search for the following:
```ts
export class DeliveryAddressItem extends defaultClasses.Base {
  @prop({ required: true })
  public name!: string;
}

export interface DeliveryAddress extends defaultClasses.Base {}
export class DeliveryAddress extends defaultClasses.TimeStamps {
  @prop({ required: true, type: () => DeliveryAddressItem, default: [] })
  public items!: mongoose.Types.DocumentArray<DocumentType<DeliveryAddressItem>>;
}
```

**Comment by xkguq007** [comment]
@hasezoey I tested simplifying it, it seemed it works! Thanks!
